integration-tests/suites/connections-and-endpoints.go

```diff
@@ -0,0 +1,177 @@
+	s.Require().NoError(err)
+	time.Sleep(6 * time.Second)
+
+	err = s.collector.TearDown()
+	s.Require().NoError(err)
+
+	s.db, err = s.collector.BoltDB()
+	s.Require().NoError(err)
+}
+
+func (s *ConnectionsAndEndpointsTestSuite) TearDownSuite() {
+	s.cleanupContainer([]string{"collector"})
+	s.cleanupContainer([]string{s.Server.Name})
+	s.cleanupContainer([]string{s.Client.Name})
+	stats := s.GetContainerStats()
+	s.PrintContainerStats(stats)
+	s.WritePerfResults("ConnectionsAndEndpoints", stats, s.metrics)
+}
+
+func (s *ConnectionsAndEndpointsTestSuite) TestConnectionsAndEndpoints() {
+
+	val, err := s.Get(s.Client.ContainerID, networkBucket)
+	s.Require().NoError(err)
+	clientNetwork, err := common.NewNetworkInfo(val)
+	s.Require().NoError(err)
+	expectedLocalAddress := strings.Replace(s.Client.ExpectedNetwork[0].LocalAddress, "CLIENT_IP", s.Client.IP, -1)
+	expectedRemoteAddress := strings.Replace(s.Client.ExpectedNetwork[0].RemoteAddress, "SERVER_IP", s.Server.IP, -1)
+	assert.Equal(s.T(), expectedLocalAddress, clientNetwork.LocalAddress)
+	assert.Equal(s.T(), expectedRemoteAddress, clientNetwork.RemoteAddress)
+	assert.Equal(s.T(), "ROLE_CLIENT", clientNetwork.Role)
+	assert.Equal(s.T(), s.Client.ExpectedNetwork[0].SocketFamily, clientNetwork.SocketFamily)
+
+	if s.Client.ExpectedEndpoints != nil {
+		fmt.Println("Expected client endpoint should be nil")
+	}
+	_, err = s.GetEndpoints(s.Client.ContainerID)
+	s.Require().Error(err, "There should be no client endpoint")
+
+	val, err = s.Get(s.Server.ContainerID, networkBucket)
+	s.Require().NoError(err)
+	serverNetwork, err := common.NewNetworkInfo(val)
+	s.Require().NoError(err)
+	expectedLocalAddress = strings.Replace(s.Server.ExpectedNetwork[0].LocalAddress, "SERVER_IP", s.Server.IP, -1)
+	expectedRemoteAddress = strings.Replace(s.Server.ExpectedNetwork[0].RemoteAddress, "CLIENT_IP", s.Client.IP, -1)
+	assert.Equal(s.T(), expectedLocalAddress, serverNetwork.LocalAddress)
+	assert.Equal(s.T(), expectedRemoteAddress, serverNetwork.RemoteAddress)
+	assert.Equal(s.T(), "ROLE_SERVER", serverNetwork.Role)
+	assert.Equal(s.T(), s.Server.ExpectedNetwork[0].SocketFamily, serverNetwork.SocketFamily)
+
+	serverEndpoints, err := s.GetEndpoints(s.Server.ContainerID)
+	s.Require().NoError(err)
+	assert.Equal(s.T(), len(s.Server.ExpectedEndpoints), len(serverEndpoints))
+
+	sort.Slice(s.Server.ExpectedEndpoints, func(i, j int) bool {
+		return endpointComparison(s.Server.ExpectedEndpoints[i], s.Server.ExpectedEndpoints[j])
+	})
+	sort.Slice(serverEndpoints, func(i, j int) bool { return endpointComparison(serverEndpoints[i], serverEndpoints[j]) })
+
+	for idx := range serverEndpoints {
+		assert.Equal(s.T(), s.Server.ExpectedEndpoints[idx].Protocol, serverEndpoints[idx].Protocol)
+		assert.Equal(s.T(), s.Server.ExpectedEndpoints[idx].Address, serverEndpoints[idx].Address)
+	}
+}
+
+func endpointComparison(endpoint1 common.EndpointInfo, endpoint2 common.EndpointInfo) bool {
```

Comment:
Is it possible to simplify this function? I'm not sure we need to compare all the fields to produce a deterministic sort order for the two lists. Perhaps something like:

```go
func endpointComparison(endpoint1, endpoint2 common.EndpointInfo) bool {
    if endpoint1.Address == nil {
        return false
    }
    if endpoint2.Address == nil {
        return true
    }
    
   addr1, addr2 := endpoint1.Address, endpoint2.Address
    return (addr1.AddressData < addr2.AddressData) && (addr1.Port < addr2.Port)
}
```

Reply:
I think we need to compare the protocol as well. What if there are two endpoints that use the same port, but different protocols? The order would not be defined. I am not sure that we need IpNetwork, so I will take it out. Also I can do the assignment 

`addr1, addr2 := endpoint1.Address, endpoint2.Address`

I think that 
`(addr1.AddressData < addr2.AddressData) && (addr1.Port < addr2.Port)`

Is not correct. What if `addr1.AddressData < addr2.AddressData` is true, but `addr1.Port < addr2.Port` is false. In that case switching `endpoint1` and `endpoint2` will also be false. The function should have the property that `endpointComparison(endpoint1, endpoint2)` returns the opposite of `endpointComparision(endpoint2, endpoint1)`, but that is not true for your suggested code change.

Reply:
yeah, the snippet I posted wasn't tested; it was just a suggestion that we could sort by address and port rather than everything. Adding Protocol into the mix makes sense, as you suggest